### PR TITLE
Support Index Operations

### PR DESCRIPTION
**Pull request by darionyaphet**

### CODE REVIEW

**Comment by jude-zhu** [comment]
Jenkins go

File: src/graph/DeleteEdgesExecutor.cpp [hunk 1]
```diff
@@ -19,30 +19,27 @@ DeleteEdgesExecutor::DeleteEdgesExecutor(Sentence *sentence,
...
-        expCtx_ = std::make_unique<ExpressionContext>();
-        expCtx_->setSpace(spaceId_);
-        expCtx_->setStorageClient(ectx()->getStorageClient());
-
-        auto edgeStatus = ectx()->schemaManager()->toEdgeType(spaceId_, *sentence_->edge());
-        if (!edgeStatus.ok()) {
-            status = edgeStatus.status();
-            break;
-        }
+    auto status = checkIfGraphSpaceChosen();
```
Comment (laura-ding) [inline src/graph/DeleteEdgesExecutor.cpp]: Everything about `space` or `schema` can't put in `prepare()`

File: src/graph/DeleteVertexExecutor.cpp [hunk 2]
```diff
@@ -16,6 +16,11 @@ DeleteVertexExecutor::DeleteVertexExecutor(Sentence *sentence,
 }
 
 Status DeleteVertexExecutor::prepare() {
+    auto status = checkIfGraphSpaceChosen();
```
Comment (laura-ding) [inline src/graph/DeleteVertexExecutor.cpp]: ditto

File: src/meta/client/MetaClient.cpp [hunk 3]
```diff
@@ -1169,6 +1169,11 @@ MetaClient::listTagIndexes(GraphSpaceID spaceID) {
     return future;
 }
 
+folly::Future<StatusOr<bool>>
+MetaClient::rebuildTagIndex(GraphSpaceID, std::string) {
+    LOG(FATAL) << "unsupported";
```
Comment (laura-ding) [inline src/meta/client/MetaClient.cpp]: Don't use LOG(FATAL), please modify to `return Status::Error("unsupported");`

File: src/meta/client/MetaClient.cpp [hunk 4]
```diff
@@ -1238,6 +1243,11 @@ MetaClient::listEdgeIndexes(GraphSpaceID spaceID) {
     return future;
 }
 
+folly::Future<StatusOr<bool>>
+MetaClient::rebuildEdgeIndex(GraphSpaceID, std::string) {
+    LOG(FATAL) << "unsupported";
```
Comment (laura-ding) [inline src/meta/client/MetaClient.cpp]: ditto

File: src/graph/ShowExecutor.cpp [hunk 5]
```diff
@@ -375,6 +381,90 @@ void ShowExecutor::showEdges() {
...
+        if (!resp.ok()) {
+            DCHECK(onError_);
+            onError_(std::move(resp).status());
+            return;
+        }
+
+        auto items = std::move(resp).value();
+        resp_ = std::make_unique<cpp2::ExecutionResponse>();
+        std::vector<cpp2::RowValue> rows;
+        std::vector<std::string> header{"Index ID", "Name"};
```
Comment (laura-ding) [inline src/graph/ShowExecutor.cpp]: The Index ID should not be known to the user, modify to `TagName`, WDYT?

File: src/graph/ShowExecutor.cpp [hunk 6]
```diff
@@ -375,6 +381,90 @@ void ShowExecutor::showEdges() {
...
```
Comment (laura-ding) [inline src/graph/ShowExecutor.cpp]: ditto

File: src/graph/ShowExecutor.cpp (hunk 5 above)
Comment (darionyaphet) [reply src/graph/ShowExecutor.cpp]: actually it's the index's name :)

**Comment by bright-starry-sky** [comment]
Please resolve the conflicts, thanks.

File: src/graph/CreateEdgeIndexExecutor.cpp [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+void CreateEdgeIndexExecutor::execute() {
+    auto status = checkIfGraphSpaceChosen();
+    if (!status.ok()) {
+        DCHECK(onError_);
+        onError_(std::move(status));
+        return;
+    }
+
+    auto *mc = ectx()->getMetaClient();
+    auto *name = sentence_->indexName();
```
Comment (bright-starry-sky) [inline src/graph/CreateEdgeIndexExecutor.cpp]: `const` is better?

File: src/graph/CreateEdgeIndexExecutor.cpp [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+    auto spaceId = ectx()->rctx()->session()->space();
+
+    std::map<std::string, std::vector<std::string>> indexProperties;
+    indexProperties.emplace(*(sentence_->edgeName()), std::move(sentence_->names()));
```
Comment (bright-starry-sky) [inline src/graph/CreateEdgeIndexExecutor.cpp]: Currently, we only support one tag or edgetype . So there is no need to use map？WDYT？

File: src/graph/CreateEdgeIndexExecutor.cpp [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+ */
+
+#include "graph/CreateEdgeIndexExecutor.h"
+
+namespace nebula {
+namespace graph {
+
+CreateEdgeIndexExecutor::CreateEdgeIndexExecutor(Sentence *sentence,
+                                                 ExecutionContext *ectx) : Executor(ectx) {
+    sentence_ = static_cast<CreateEdgeIndexSentence*>(sentence);
```
Comment (bright-starry-sky) [inline src/graph/CreateEdgeIndexExecutor.cpp]: `dynamic_cast` is better？

File: src/graph/CreateEdgeIndexExecutor.cpp (hunk 8 above)
Comment (bright-starry-sky) [inline src/graph/CreateEdgeIndexExecutor.cpp]: Don't need `std::move `

File: src/graph/CreateEdgeIndexExecutor.h [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+        return "CreateEdgeIndexExecutor";
+    }
+
+    Status MUST_USE_RESULT prepare() override;
+
+    void execute() override;
+
+private:
+    CreateEdgeIndexSentence       *sentence_{nullptr};
+    const std::string             *indexName_{nullptr};
```
Comment (bright-starry-sky) [inline src/graph/CreateEdgeIndexExecutor.h]: 'indexName_' is never used.

File: src/graph/CreateTagIndexExecutor.cpp [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+        DCHECK(onError_);
+        onError_(std::move(status));
+        return;
+    }
+
+    auto *mc = ectx()->getMetaClient();
+    auto *name = sentence_->indexName();
+    auto spaceId = ectx()->rctx()->session()->space();
+
+    std::map<std::string, std::vector<std::string>> indexProperties;
```
Comment (bright-starry-sky) [inline src/graph/CreateTagIndexExecutor.cpp]: Need to use a map?

File: src/graph/CreateTagIndexExecutor.cpp [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+void CreateTagIndexExecutor::execute() {
+    auto status = checkIfGraphSpaceChosen();
+    if (!status.ok()) {
+        DCHECK(onError_);
+        onError_(std::move(status));
+        return;
+    }
+
+    auto *mc = ectx()->getMetaClient();
+    auto *name = sentence_->indexName();
```
Comment (bright-starry-sky) [inline src/graph/CreateTagIndexExecutor.cpp]: please use `indexName_`

File: src/graph/CreateTagIndexExecutor.cpp [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+    auto spaceId = ectx()->rctx()->session()->space();
+
+    std::map<std::string, std::vector<std::string>> indexProperties;
+    indexProperties.emplace(*(sentence_->tagName()), std::move(sentence_->names()));
```
Comment (bright-starry-sky) [inline src/graph/CreateTagIndexExecutor.cpp]: std::move

File: src/graph/DeleteEdgesExecutor.cpp [hunk 14]
```diff
@@ -19,34 +19,34 @@ DeleteEdgesExecutor::DeleteEdgesExecutor(Sentence *sentence,
...
-        expCtx_ = std::make_unique<ExpressionContext>();
-        expCtx_->setSpace(spaceId_);
-        expCtx_->setStorageClient(ectx()->getStorageClient());
-
-        auto edgeStatus = ectx()->schemaManager()->toEdgeType(spaceId_, *sentence_->edge());
-        if (!edgeStatus.ok()) {
-            status = edgeStatus.status();
-            break;
-        }
+    spaceId_ = ectx()->rctx()->session()->space();
```
Comment (bright-starry-sky) [inline src/graph/DeleteEdgesExecutor.cpp]: What's reason move the `checkIfGraphSpaceChosen` to `execute()`?

File: src/graph/DeleteVertexExecutor.cpp [hunk 15]
```diff
@@ -33,7 +33,13 @@ Status DeleteVertexExecutor::prepare() {
 }
 
 void DeleteVertexExecutor::execute() {
-    // TODO(zlcook) Get edgeKeys of a vertex by Go
+    auto status = checkIfGraphSpaceChosen();
```
Comment (bright-starry-sky) [inline src/graph/DeleteVertexExecutor.cpp]: ditto

File: src/graph/DescribeEdgeIndexExecutor.cpp [hunk 16]
```diff
@@ -0,0 +1,82 @@
...
+
+void DescribeEdgeIndexExecutor::execute() {
+    auto status = checkIfGraphSpaceChosen();
+    if (!status.ok()) {
+        DCHECK(onError_);
+        onError_(std::move(status));
+        return;
+    }
+
+    auto *name = sentence_->indexName();
```
Comment (bright-starry-sky) [inline src/graph/DescribeEdgeIndexExecutor.cpp]: const is better?

File: src/graph/RebuildEdgeIndexExecutor.cpp [hunk 17]
```diff
@@ -0,0 +1,55 @@
...
+    auto status = checkIfGraphSpaceChosen();
+    if (!status.ok()) {
+        DCHECK(onError_);
+        onError_(std::move(status));
+        return;
+    }
+
+    auto *mc = ectx()->getMetaClient();
+    auto *name = sentence_->indexName();
+    auto spaceId = ectx()->rctx()->session()->space();
```
Comment (bright-starry-sky) [inline src/graph/RebuildEdgeIndexExecutor.cpp]: Could we move the rebuild index to another PR?

File: src/graph/SchemaHelper.cpp [hunk 18]
```diff
@@ -49,24 +49,21 @@ Status SchemaHelper::createSchema(const std::vector<ColumnSpecification*>& specs
...
                     break;
                 case nebula::ColumnType::STRING:
                     v.set_string_value(spec->getStringValue());
-                    column.set_default_value(std::move(v));
                     break;
                 default:
                     LOG(ERROR) << "Unsupport Type";
                     return Status::Error("Unsupport Type");
             }
+            column.set_default_value(std::move(v));
```
Comment (bright-starry-sky) [inline src/graph/SchemaHelper.cpp]: 🐂

File: src/graph/ShowExecutor.cpp [hunk 19]
```diff
@@ -375,6 +381,90 @@ void ShowExecutor::showEdges() {
...
+void ShowExecutor::showTagIndexes() {
+    auto spaceId = ectx()->rctx()->session()->space();
+    auto future = ectx()->getMetaClient()->listTagIndexes(spaceId);
+    auto *runner = ectx()->rctx()->runner();
+    resp_ = std::make_unique<cpp2::ExecutionResponse>();
+
+    auto cb = [this] (auto &&resp) {
+        if (!resp.ok()) {
+            DCHECK(onError_);
+            onError_(std::move(resp).status());
```
Comment (bright-starry-sky) [inline src/graph/ShowExecutor.cpp]: `onError_(std::forward<decltype(resp)>(resp).status());` is better?

File: src/graph/ShowExecutor.cpp [hunk 20]
```diff
@@ -375,6 +381,90 @@ void ShowExecutor::showEdges() {
...
+            return;
+        }
+
+        auto items = std::move(resp).value();
```
Comment (bright-starry-sky) [inline src/graph/ShowExecutor.cpp]: `auto items = std::forward<decltype(resp)>(resp).value(); `is better?

File: src/graph/ShowExecutor.cpp [hunk 21]
```diff
@@ -375,6 +381,90 @@ void ShowExecutor::showEdges() {
...
+void ShowExecutor::showEdgeIndexes() {
+    auto spaceId = ectx()->rctx()->session()->space();
+    auto future = ectx()->getMetaClient()->listEdgeIndexes(spaceId);
+    auto *runner = ectx()->rctx()->runner();
+    resp_ = std::make_unique<cpp2::ExecutionResponse>();
+
+    auto cb = [this] (auto &&resp) {
+        if (!resp.ok()) {
+            DCHECK(onError_);
+            onError_(std::move(resp).status());
```
Comment (bright-starry-sky) [inline src/graph/ShowExecutor.cpp]: std::forward

File: src/graph/ShowExecutor.cpp [hunk 22]
```diff
@@ -375,6 +381,90 @@ void ShowExecutor::showEdges() {
...
+            return;
+        }
+
+        auto items = std::move(resp).value();
```
Comment (bright-starry-sky) [inline src/graph/ShowExecutor.cpp]: std::forward

File: src/graph/ShowExecutor.cpp [hunk 23]
```diff
@@ -53,6 +51,12 @@ void ShowExecutor::execute() {
         case ShowSentence::ShowType::kShowEdges:
             showEdges();
             break;
+        case ShowSentence::ShowType::kShowTagIndexes:
+            showTagIndexes();
+            break;
+        case ShowSentence::ShowType::kShowEdgeIndexes:
+            showEdgeIndexes();
+            break;
```
Comment (bright-starry-sky) [inline src/graph/ShowExecutor.cpp]: could we provide a method for show index? for example : 
show tag index(name)
indexID | indexName | col1 | col2 | col3 | ....

Comment (bright-starry-sky) [reply src/graph/ShowExecutor.cpp]: Aha , showCreateTagIndex and showCreateEdgeIndex seems to have the same purpose

File: src/parser/MaintainSentences.cpp [hunk 24]
```diff
@@ -201,4 +201,61 @@ std::string DropTagSentence::toString() const {
...
+
+std::string CreateTagIndexSentence::toString() const {
+    std::string buf;
+    buf.reserve(256);
+    buf += "CREATE TAG INDEX ";
+    buf += *indexName_;
+    buf += " ON ";
+    buf += *tagName_;
+    buf += " (";
+    buf += ")";
```
Comment (bright-starry-sky) [inline src/parser/MaintainSentences.cpp]: column name?

File: src/parser/MaintainSentences.cpp [hunk 25]
```diff
@@ -201,4 +201,61 @@ std::string DropTagSentence::toString() const {
...
+
+std::string CreateEdgeIndexSentence::toString() const {
+    std::string buf;
+    buf.reserve(256);
+    buf += "CREATE EDGE INDEX ";
+    buf += *indexName_;
+    buf += " ON ";
+    buf += *edgeName_;
+    buf += " (";
+    buf += ")";
```
Comment (bright-starry-sky) [inline src/parser/MaintainSentences.cpp]: ditto

File: src/parser/MaintainSentences.h [hunk 26]
```diff
@@ -496,6 +497,201 @@ class DropEdgeSentence final : public Sentence {
...
+
+    const std::string* indexName() const {
+        return indexName_.get();
+    }
+
+    const std::string* tagName() const {
+        return tagName_.get();
+    }
+
+    const ColumnNameList* columnNames() const {
```
Comment (bright-starry-sky) [inline src/parser/MaintainSentences.h]: 'columnNames' is never used

File: src/parser/MaintainSentences.h [hunk 27]
```diff
@@ -496,6 +497,201 @@ class DropEdgeSentence final : public Sentence {
...
+
+    const std::string* indexName() const {
+        return indexName_.get();
+    }
+
+    const std::string* edgeName() const {
+        return edgeName_.get();
+    }
+
+    const ColumnNameList* columnNames() const {
```
Comment (bright-starry-sky) [inline src/parser/MaintainSentences.h]: 'columnNames' is never used

File: src/parser/parser.yy [hunk 28]
```diff
@@ -99,20 +99,20 @@ class GraphScanner;
...
+%token KW_VARIABLES KW_GET KW_DECLARE KW_GRAPH KW_META KW_STORAGE
+%token KW_TTL KW_TTL_DURATION KW_TTL_COL KW_DATA KW_STOP
 %token KW_FETCH KW_PROP KW_UPDATE KW_UPSERT KW_WHEN
+%token KW_ORDER KW_ASC KW_LIMIT KW_OFFSET KW_GROUP
 %token KW_DISTINCT KW_ALL KW_OF
-%token KW_BALANCE KW_LEADER KW_DATA KW_STOP
+%token KW_BALANCE KW_LEADER
 %token KW_SHORTEST KW_PATH
-%token KW_IS KW_NULL
+%token KW_IS KW_NULL KW_DEFAULT
```
Comment (bright-starry-sky) [inline src/parser/parser.yy]: One small suggestion，Please don't change the order easily，This can lead to confusion for others about merging code . 😄

File: src/parser/parser.yy [hunk 29]
```diff
@@ -198,20 +198,26 @@ class GraphScanner;
 %type <sentence> go_sentence match_sentence use_sentence find_sentence find_path_sentence
 %type <sentence> order_by_sentence limit_sentence group_by_sentence
 %type <sentence> fetch_vertices_sentence fetch_edges_sentence
+
```
Comment (bright-starry-sky) [inline src/parser/parser.yy]: ditto

File: src/parser/parser.yy [hunk 30]
```diff
@@ -1115,6 +1121,60 @@ drop_edge_sentence
     }
     ;
 
+create_tag_index_sentence
+    : KW_CREATE KW_TAG KW_INDEX opt_if_not_exists name_label KW_ON name_label L_PAREN column_name_list R_PAREN {
```
Comment (bright-starry-sky) [inline src/parser/parser.yy]: `opt_if_not_exists` is available now?  I suggest deleting it if not yet.

File: src/parser/parser.yy [hunk 31]
```diff
@@ -1115,6 +1121,60 @@ drop_edge_sentence
...
+        $$ = new CreateTagIndexSentence($5, $7, $9, $4);
+    }
+    ;
+
+create_edge_index_sentence
+    : KW_CREATE KW_EDGE KW_INDEX opt_if_not_exists name_label KW_ON name_label L_PAREN column_name_list R_PAREN {
```
Comment (bright-starry-sky) [inline src/parser/parser.yy]: ditto

File: src/parser/parser.yy [hunk 32]
```diff
@@ -1510,7 +1582,7 @@ show_config_item
     ;
 
 create_space_sentence
-    : KW_CREATE KW_SPACE opt_if_not_exists  name_label{
+    : KW_CREATE KW_SPACE opt_if_not_exists name_label {
```
Comment (bright-starry-sky) [inline src/parser/parser.yy]: good !

File: src/graph/CreateEdgeIndexExecutor.cpp (hunk 7 above)
Comment (darionyaphet) [reply src/graph/CreateEdgeIndexExecutor.cpp]: I don't know what's the practical use ?

Comment (bright-starry-sky) [reply src/graph/CreateEdgeIndexExecutor.cpp]: > I don't know what's the practical use ?
Actually it's just my code like, here's what I think ：
Reason 1 : The method `sentence_->indexName()` return value is a const string *. 
Reason 2 : This value is an immutable constant.

File: src/graph/CreateEdgeIndexExecutor.cpp (hunk 8 above)
Comment (darionyaphet) [reply src/graph/CreateEdgeIndexExecutor.cpp]: emmm... this is a long story. Long long ago, we design the index will across multi tag or edge.

File: src/graph/CreateEdgeIndexExecutor.cpp (hunk 7 above)
Comment (darionyaphet) [reply src/graph/CreateEdgeIndexExecutor.cpp]: I think the index name is immutable. The compiler maybe have some optimization for the `const auto*` .

File: src/graph/DeleteEdgesExecutor.cpp (hunk 14 above)
Comment (darionyaphet) [reply src/graph/DeleteEdgesExecutor.cpp]: For example : use space; delete index ...

File: src/graph/ShowExecutor.cpp (hunk 21 above)
Comment (darionyaphet) [reply src/graph/ShowExecutor.cpp]: `std::move` takes an object and allows you to treat it as a temporary. 

`std::forward` has a single use case: to cast a templated function parameter to the value category the caller used to pass it. 

When the parameter is an rvalue, they are treated are the same.

File: src/graph/CreateEdgeIndexExecutor.cpp (hunk 8 above)
Comment (darionyaphet) [reply src/graph/CreateEdgeIndexExecutor.cpp]: why ?

File: src/graph/CreateEdgeIndexExecutor.cpp (hunk 9 above)
Comment (darionyaphet) [reply src/graph/CreateEdgeIndexExecutor.cpp]: when using `dynamic_cast` transfer derived to base class, it's more safety than `static_cast`. Because the sentence is extend from `Sentence` when it declared so I think it's ok at here.

File: src/graph/DescribeEdgeIndexExecutor.cpp [hunk 33]
```diff
@@ -0,0 +1,82 @@
...
+                                           ExecutionContext *ectx) : Executor(ectx) {
+    sentence_ = static_cast<DescribeEdgeIndexSentence*>(sentence);
+}
+
+Status DescribeEdgeIndexExecutor::prepare() {
+    return Status::OK();
+}
+
+
+void DescribeEdgeIndexExecutor::execute() {
```
Comment (critical27) [inline src/graph/DescribeEdgeIndexExecutor.cpp]: Do we need desc for index, `show index` is enough?

File: src/meta/client/MetaClient.cpp [hunk 34]
```diff
@@ -1211,6 +1236,20 @@ MetaClient::createEdgeIndex(GraphSpaceID spaceID,
...
+MetaClient::createEdgeIndex(GraphSpaceID spaceID,
+                            std::string indexName,
+                            std::string edgeName,
+                            std::vector<std::string> fields,
+                            bool ifNotExists) {
+    std::map<std::string, std::vector<std::string>> properties;
+    properties.emplace(std::move(edgeName), std::move(fields));
+    return createEdgeIndex(spaceID,
+                           std::move(indexName),
+                           std::move(properties),
```
[the remark this answers is not in the document]
Comment (darionyaphet) [reply src/meta/client/MetaClient.cpp]: actually I have update to single schema in another PR

File: src/graph/DescribeEdgeIndexExecutor.cpp (hunk 33 above)
Comment (darionyaphet) [reply src/graph/DescribeEdgeIndexExecutor.cpp]: `SHOW TAG/EDGE INDEXES` will display an index list.

File: src/parser/scanner.lex [hunk 35]
```diff
@@ -188,6 +192,10 @@ IP_OCTET                    ([0-9]|[1-9][0-9]|1[0-9][0-9]|2[0-4][0-9]|25[0-5])
 {REVERSELY}                 { return TokenType::KW_REVERSELY; }
 {SPACE}                     { return TokenType::KW_SPACE; }
 {SPACES}                    { return TokenType::KW_SPACES; }
+{INDEX}                     { return TokenType::KW_INDEX; }
+{INDEXES}                   { return TokenType::KW_INDEXES; }
+{BUILD}                     { return TokenType::KW_BUILD; }
+{TTL}                       { return TokenType::KW_TTL; }
```
Comment (bright-starry-sky) [inline src/parser/scanner.lex]: What purpose of `TTL`?

File: src/meta/processors/indexMan/CreateEdgeIndexProcessor.cpp [hunk 36]
```diff
@@ -25,7 +25,11 @@ void CreateEdgeIndexProcessor::process(const cpp2::CreateEdgeIndexReq& req) {
     auto ret = getEdgeIndexID(space, indexName);
     if (ret.ok()) {
         LOG(ERROR) << "Create Edge Index Failed: " << indexName << " have existed";
-        resp_.set_code(cpp2::ErrorCode::E_EXISTED);
+        if (req.get_if_not_exists()) {
+            resp_.set_code(cpp2::ErrorCode::SUCCEEDED);
+        } else {
+            resp_.set_code(cpp2::ErrorCode::E_EXISTED);
+        }
```
[the remark this answers is not in the document]
Comment (darionyaphet) [reply src/meta/processors/indexMan/CreateEdgeIndexProcessor.cpp]: Whether or not the `if_not_exists` is setting the create index is failed.

[the remark this answers is not in the document]
Comment (bright-starry-sky) [reply src/meta/processors/indexMan/CreateEdgeIndexProcessor.cpp]: If the if_not_exists == true , no error message output is required.

File: src/meta/processors/indexMan/CreateTagIndexProcessor.cpp [hunk 37]
```diff
@@ -25,7 +25,11 @@ void CreateTagIndexProcessor::process(const cpp2::CreateTagIndexReq& req) {
     auto ret = getTagIndexID(space, indexName);
     if (ret.ok()) {
         LOG(ERROR) << "Create Tag Index Failed: " << indexName << " have existed";
-        resp_.set_code(cpp2::ErrorCode::E_EXISTED);
+        if (req.get_if_not_exists()) {
+            resp_.set_code(cpp2::ErrorCode::SUCCEEDED);
+        } else {
+            resp_.set_code(cpp2::ErrorCode::E_EXISTED);
+        }
```
Comment (bright-starry-sky) [inline src/meta/processors/indexMan/CreateTagIndexProcessor.cpp]: ditto

**Comment by darionyaphet** [comment]
Reference document : #1578

File: src/meta/processors/indexMan/CreateEdgeIndexProcessor.cpp (hunk 36 above)
[the remark this answers is not in the document]
Comment (bright-starry-sky) [reply src/meta/processors/indexMan/CreateEdgeIndexProcessor.cpp]: `LOG(ERROR) << "Create Edge Index Failed: " << indexName << " have existed";`
这行代码，我认为只在if_not_exists == false的时候才输出。

**Comment by jude-zhu** [comment]
close #470

**Comment by jude-zhu** [comment]
relate to #460
